### PR TITLE
[bugfix] Re-compute accumulated_grad_batches

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds logic to recompute `accumulated_grad_batches` before `on_train_start`.

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/5334

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/gradient_accumulation_scheduler.py [hunk 1]
```diff
@@ -74,9 +74,16 @@ def __init__(self, scheduling: Dict[int, int]):
     def going_to_accumulate_grad_batches(self):
         return any(v > 1 for v in self.scheduling.values())
 
-    def on_train_epoch_start(self, trainer, pl_module):
-        epoch = trainer.current_epoch
+    def _get_current_accumulate_grad_batches(self, epoch: int) -> int:
```
Comment (justusschock) [inline pytorch_lightning/callbacks/gradient_accumulation_scheduler.py:77]: why do we need to split this? can't we just inline this to `on_train_epoch_start`? Same for `compute_accumulate_grad_batches. IMO this should all be inlined since this seems to be the only place this is called.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/gradient_accumulation_scheduler.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/gradient_accumulation_scheduler.py]: Removed the extra function.

Comment (tchaton) [reply pytorch_lightning/callbacks/gradient_accumulation_scheduler.py]: This is required as we need to compute the current `accumulate_grad_batches` and calling `on_train_epoch_start` there could be confusing for readers.

File: tests/checkpointing/test_trainer_checkpoint.py [hunk 2]
```diff
@@ -84,3 +85,25 @@ def validation_step(self, batch, batch_idx):
...
+    trainer_kwargs = dict(
+        max_epochs=1, accumulate_grad_batches={0: 2}, callbacks=cb, limit_train_batches=1, limit_val_batches=0
+    )
+    trainer = Trainer(**trainer_kwargs)
+    trainer.fit(model, train_data, val_data)
+
+    trainer_kwargs['max_epochs'] = 2
+    trainer_kwargs['resume_from_checkpoint'] = cb.last_model_path
+    trainer = Trainer(**trainer_kwargs)
+    trainer.fit(model, train_data, val_data)
```
Comment (awaelchli) [inline tests/checkpointing/test_trainer_checkpoint.py:109]:
```suggestion
def test_accumulated_gradient_batches_with_resume_from_checkpoint(tmpdir):
    """
    This test validates that accumulated gradient properly is recomputed and reset on the trainer.
    """
    cb = ModelCheckpoint(dirpath=tmpdir, save_last=True)
    model = BoringModel()
    trainer_kwargs = dict(
        max_epochs=1, accumulate_grad_batches={0: 2}, callbacks=cb, limit_train_batches=1, limit_val_batches=0
    )
    trainer = Trainer(**trainer_kwargs)
    trainer.fit(model)

    trainer_kwargs['max_epochs'] = 2
    trainer_kwargs['resume_from_checkpoint'] = cb.last_model_path
    trainer = Trainer(**trainer_kwargs)
    trainer.fit(model)
```

Comment (awaelchli) [reply tests/checkpointing/test_trainer_checkpoint.py:109]: The boringmodel already has dataloaders defined

**Review by awaelchli — APPROVED** [review]

File: tests/checkpointing/test_trainer_checkpoint.py (hunk 2 above)
Comment (tchaton) [reply tests/checkpointing/test_trainer_checkpoint.py:109]: Removed !

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by kaushikb11 — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
